src/librustc_mir_build/build/expr/as_operand.rs

```diff
@@ -21,6 +21,37 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         self.as_operand(block, local_scope, expr)
     }
 
+    /// Returns an operand suitable for use until the end of the current
+    /// scope expression and suitable also to be passed as function arguments.
+    ///
+    /// The operand returned from this function will *not be valid* after
+    /// an ExprKind::Scope is passed, so please do *not* return it from
+    /// functions to avoid bad miscompiles.
+    crate fn as_call_temporary<M>(&mut self, block: BasicBlock, expr: M) -> BlockAnd<Operand<'tcx>>
+    where
+        M: Mirror<'tcx, Output = Expr<'tcx>>,
+    {
+        let local_scope = self.local_scope();
+        let expr = self.hir.mirror(expr);
+
+        if let ExprKind::Deref { ref arg } = expr.kind {
+            let tcx = self.hir.tcx();
+            let param_env = self.hir.param_env;
+
+            let arg = self.hir.mirror(arg.clone());
+            let ty = arg.ty;
+            let span = arg.span;
+
+            if !ty.is_sized(tcx.at(span), param_env)
+                && !ty.is_copy_modulo_regions(tcx, param_env, span)
+            {
+                return self.expr_as_operand(block, local_scope, arg);
```

Comment:
So I think what this should be is invoke `as_temp` to get back a `Local` (let's call it `tmp`), and then return the operand of `Operand::Move(*tmp)`.